core/src/commonMain/kotlin/com/powersync/db/internal/InternalDatabaseImpl.kt

```diff
@@ -189,21 +160,31 @@ internal class InternalDatabaseImpl(
     override fun <RowType : Any> watch(
         sql: String,
         parameters: List<Any?>?,
+        throttleMs: Long?,
         mapper: (SqlCursor) -> RowType,
-    ): Flow<List<RowType>> {
-        val tables =
-            getSourceTables(sql, parameters)
-                .map { toFriendlyTableName(it) }
-                .filter { it.isNotBlank() }
-                .toSet()
-        return watchQuery(
-            query = sql,
-            parameters = parameters?.size ?: 0,
-            binders = getBindersFromParams(parameters),
-            mapper = mapper,
-            tables = tables,
-        ).asFlow().mapToList(scope.coroutineContext)
-    }
+    ): Flow<List<RowType>> =
+        // Use a channel flow here since we throttle (buffer used under the hood)
+        // This causes some emissions to be from different scopes.
+        channelFlow {
+            // Fetch the tables asynchronously with getAll
+            val tables =
+                getSourceTables(sql, parameters)
+                    .filter { it.isNotBlank() }
+                    .toSet()
+
+            updatesOnTables()
+                // onSubscription here is very important.
+                // This ensures that the initial result and all updates are emitted.
+                .onSubscription {
+                    send(getAll(sql, parameters = parameters, mapper = mapper))
+                }.filter {
+                    // Only trigger updates on relevant tables
+                    it.intersect(tables).isNotEmpty()
+                }.throttle(throttleMs ?: DEFAULT_WATCH_THROTTLE_MS)
```

Comment:
Maybe add a comment here why `throttle` is fine despite possibly dropping events when `getAll` is too slow (in my understanding, it's because we only need to ensure that an update after `getAll` started will cause another emission later, and that's guaranteed because we're keeping the latest event).

Reply:
Good point. I added a comment.